commit 92bfa3fd49903057a903ddfa82c92d66ae16a607
Author:     Nicholas Wilson <nicholas@nicholaswilson.me.uk>
AuthorDate: Thu Aug 27 16:52:16 2026 +0100
Commit:     Nicholas Wilson <nicholas@nicholaswilson.me.uk>
CommitDate: Sat Aug 29 21:18:43 2026 +0100

    Backport of pcre2-10.48-Fix-invalid-UTF-backwards-scans.patch
    
    Cherry-pick of 4889caf31a4c5a6b3c051f0031bf2dbd78f2c287
    
    Fix invalid UTF backwards-scan reads; see GHSA-9qww-pwc4-77qq for details

diff --git a/src/pcre2_extuni.c b/src/pcre2_extuni.c
index b23946b..b0e7335 100644
--- a/src/pcre2_extuni.c
+++ b/src/pcre2_extuni.c
@@ -57,12 +57,12 @@ support, because some compilers do not like functionless source files. */
 
 #ifndef SUPPORT_UNICODE
 PCRE2_SPTR
-PRIV(extuni)(uint32_t c, PCRE2_SPTR eptr, PCRE2_SPTR start_subject,
+PRIV(extuni)(uint32_t c, PCRE2_SPTR eptr, PCRE2_SPTR check_subject,
   PCRE2_SPTR end_subject, BOOL utf, int *xcount)
 {
 (void)c;
 (void)eptr;
-(void)start_subject;
+(void)check_subject;
 (void)end_subject;
 (void)utf;
 (void)xcount;
@@ -79,7 +79,7 @@ return NULL;
 Arguments:
   c              the first character
   eptr           pointer to next character
-  start_subject  pointer to start of subject
+  check_subject  pointer to start of validated subject
   end_subject    pointer to end of subject
   utf            TRUE if in UTF mode
   xcount         pointer to count of additional characters,
@@ -89,7 +89,7 @@ Returns:         pointer after the end of the sequence
 */
 
 PCRE2_SPTR
-PRIV(extuni)(uint32_t c, PCRE2_SPTR eptr, PCRE2_SPTR start_subject,
+PRIV(extuni)(uint32_t c, PCRE2_SPTR eptr, PCRE2_SPTR check_subject,
   PCRE2_SPTR end_subject, BOOL utf, int *xcount)
 {
 int lgb = UCD_GRAPHBREAK(c);
@@ -113,7 +113,7 @@ while (eptr < end_subject)
 
     /* bptr is pointing to the left-hand character */
 
-    while (bptr > start_subject)
+    while (bptr > check_subject)
       {
       bptr--;
       if (utf)
diff --git a/src/pcre2_match.c b/src/pcre2_match.c
index 83ab2bc..68e9147 100644
--- a/src/pcre2_match.c
+++ b/src/pcre2_match.c
@@ -2579,7 +2579,7 @@ fprintf(stderr, "++ op=%d\n", *Fecode);
     else
       {
       GETCHARINCTEST(fc, Feptr);
-      Feptr = PRIV(extuni)(fc, Feptr, mb->start_subject, mb->end_subject, utf,
+      Feptr = PRIV(extuni)(fc, Feptr, mb->check_subject, mb->end_subject, utf,
         NULL);
       }
     CHECK_PARTIAL();
@@ -2932,7 +2932,7 @@ fprintf(stderr, "++ op=%d\n", *Fecode);
           else
             {
             GETCHARINCTEST(fc, Feptr);
-            Feptr = PRIV(extuni)(fc, Feptr, mb->start_subject,
+            Feptr = PRIV(extuni)(fc, Feptr, mb->check_subject,
               mb->end_subject, utf, NULL);
             }
           CHECK_PARTIAL();
@@ -3754,7 +3754,7 @@ fprintf(stderr, "++ op=%d\n", *Fecode);
           else
             {
             GETCHARINCTEST(fc, Feptr);
-            Feptr = PRIV(extuni)(fc, Feptr, mb->start_subject, mb->end_subject,
+            Feptr = PRIV(extuni)(fc, Feptr, mb->check_subject, mb->end_subject,
               utf, NULL);
             }
           CHECK_PARTIAL();
@@ -4335,7 +4335,7 @@ fprintf(stderr, "++ op=%d\n", *Fecode);
           else
             {
             GETCHARINCTEST(fc, Feptr);
-            Feptr = PRIV(extuni)(fc, Feptr, mb->start_subject, mb->end_subject,
+            Feptr = PRIV(extuni)(fc, Feptr, mb->check_subject, mb->end_subject,
               utf, NULL);
             }
           CHECK_PARTIAL();
diff --git a/testdata/testinput10 b/testdata/testinput10
index 53e37cb..ef9c311 100644
--- a/testdata/testinput10
+++ b/testdata/testinput10
@@ -555,6 +555,12 @@
     AAA\x80BXYZ 
     AAA\x80BBXYZ 
 
+# Backward scans must not enter the invalid prefix before check_subject.
+
+/\X\X/utf,match_invalid_utf
+\= Expect no match
+    \x80\x{1f1e6}\x{1f1e7}
+
 # -------------------------------------
 
 /(*UTF)(?=\x{123})/I
diff --git a/testdata/testinput12 b/testdata/testinput12
index 9b4f8d3..f526808 100644
--- a/testdata/testinput12
+++ b/testdata/testinput12
@@ -451,6 +451,12 @@
 /aa/utf,ucp,match_invalid_utf,global
     \x{d800}aa
 
+# Backward scans must not enter the invalid prefix before check_subject.
+
+/\X\X/utf,match_invalid_utf
+\= Expect no match
+    \x{dc00}\x{1f1e6}\x{1f1e7}
+
 # ---------------------------------------------------- 
 
 /(*UTF)(?=\x{123})/I
diff --git a/testdata/testoutput10 b/testdata/testoutput10
index d408510..893e357 100644
--- a/testdata/testoutput10
+++ b/testdata/testoutput10
@@ -1727,6 +1727,13 @@ No match
     AAA\x80BBXYZ 
 No match
 
+# Backward scans must not enter the invalid prefix before check_subject.
+
+/\X\X/utf,match_invalid_utf
+\= Expect no match
+    \x80\x{1f1e6}\x{1f1e7}
+No match
+
 # -------------------------------------
 
 /(*UTF)(?=\x{123})/I
diff --git a/testdata/testoutput12-16 b/testdata/testoutput12-16
index 84c4858..df69365 100644
--- a/testdata/testoutput12-16
+++ b/testdata/testoutput12-16
@@ -1586,6 +1586,13 @@ No match
     \x{d800}aa
  0: aa
 
+# Backward scans must not enter the invalid prefix before check_subject.
+
+/\X\X/utf,match_invalid_utf
+\= Expect no match
+    \x{dc00}\x{1f1e6}\x{1f1e7}
+No match
+
 # ---------------------------------------------------- 
 
 /(*UTF)(?=\x{123})/I
diff --git a/testdata/testoutput12-32 b/testdata/testoutput12-32
index 03b6e39..c47c832 100644
--- a/testdata/testoutput12-32
+++ b/testdata/testoutput12-32
@@ -1584,6 +1584,13 @@ No match
     \x{d800}aa
  0: aa
 
+# Backward scans must not enter the invalid prefix before check_subject.
+
+/\X\X/utf,match_invalid_utf
+\= Expect no match
+    \x{dc00}\x{1f1e6}\x{1f1e7}
+No match
+
 # ---------------------------------------------------- 
 
 /(*UTF)(?=\x{123})/I
